android/src/main/kotlin/com/example/flutter_gamepad/FlutterGamepadPlugin.kt

```diff
@@ -1,25 +1,91 @@
 package com.example.flutter_gamepad
 
+import android.view.KeyEvent
+import android.view.MotionEvent
+import io.flutter.embedding.android.AndroidKeyProcessor
+import io.flutter.embedding.android.AndroidTouchProcessor
+import io.flutter.view.FlutterView
+import io.flutter.embedding.engine.renderer.FlutterRenderer
+import io.flutter.embedding.engine.systemchannels.KeyEventChannel
+import io.flutter.plugin.common.EventChannel
 import io.flutter.plugin.common.MethodCall
 import io.flutter.plugin.common.MethodChannel
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler
 import io.flutter.plugin.common.MethodChannel.Result
 import io.flutter.plugin.common.PluginRegistry.Registrar
+import io.flutter.plugin.editing.TextInputPlugin
+import java.lang.reflect.Field
 
-class FlutterGamepadPlugin: MethodCallHandler {
-  companion object {
-    @JvmStatic
-    fun registerWith(registrar: Registrar) {
-      val channel = MethodChannel(registrar.messenger(), "flutter_gamepad")
-      channel.setMethodCallHandler(FlutterGamepadPlugin())
+/**
+ * An extension of Flutter's AndroidTouchProcessor that delegates MotionEvents to the GamepadStreamHandler.
+ */
+class GamepadAndroidTouchProcessor(renderer: FlutterRenderer) : AndroidTouchProcessor(renderer) {
+    override fun onGenericMotionEvent(event: MotionEvent): Boolean {
+        return GamepadStreamHandler.processMotionEvent(event) || super.onGenericMotionEvent(event)
+    }
+}
+
+/**
+ * An extension of Flutter's AndroidKeyProcessor that delegates KeyEvents to the GamepadStreamHandler.
+ */
+class GamepadAndroidKeyProcessor(keyEventChannel: KeyEventChannel, textInputPlugin: TextInputPlugin) : AndroidKeyProcessor(keyEventChannel, textInputPlugin) {
+    override fun onKeyDown(keyEvent: KeyEvent) {
+        val handled = GamepadStreamHandler.processKeyDownEvent(keyEvent)
+        if (!handled) {
+            super.onKeyDown(keyEvent)
+        }
+    }
+
+    override fun onKeyUp(keyEvent: KeyEvent) {
+        val handled = GamepadStreamHandler.processKeyUpEvent(keyEvent)
+        if (!handled) {
+            super.onKeyDown(keyEvent)
+        }
+    }
+}
+
+/**
+ * The flutter_gamepad plugin class that is registered with the framework.
+ */
+class FlutterGamepadPlugin : MethodCallHandler {
+    companion object {
+        @JvmStatic
+        fun registerWith(registrar: Registrar) {
+            val methodChannel = MethodChannel(registrar.messenger(), "com.rainway.flutter_gamepad/methods")
+            methodChannel.setMethodCallHandler(FlutterGamepadPlugin())
+            val eventChannel = EventChannel(registrar.messenger(), "com.rainway.flutter_gamepad/events")
+            eventChannel.setStreamHandler(GamepadStreamHandler)
+
+            val view: FlutterView = registrar.view()
+            fun viewField(name: String): Field {
+                val field = FlutterView::class.java.getDeclaredField(name)
```

Comment:
Does this need a try/catch?

Reply:
See below: if this process goes wrong for someone after a Flutter-internal change, we can't really save anything and probably want to let that error propagate so they can tell us it broke.